source/sigv4.c

```diff
@@ -437,6 +562,398 @@ static SigV4Status_t parseDate( const char * pDate,
+            lenSmall = pFirstVal->key.dataLen;
+        }
+        else
+        {
+            lenSmall = pSecondVal->key.dataLen;
+        }
+
+        return strncmp( ( char * ) pFirstVal->key.pData,
+                        ( char * ) pSecondVal->key.pData,
+                        lenSmall );
+    }
+
+/*-----------------------------------------------------------*/
+
+    static void encodeURI( const char * pURI,
+                           size_t uriLen,
+                           char * pCanonicalURI,
+                           size_t * canonicalURILen,
+                           bool encodeSlash,
+                           bool nullTerminate )
+    {
+        const char * pURILoc = pURI;
+        char * pBufLoc = pCanonicalURI;
+        size_t index = 0U;
+
+        assert( pURI != NULL );
+        assert( pCanonicalURI != NULL );
+        assert( canonicalURILen != NULL );
+        assert( *canonicalURILen > 0U );
+
+        while( index < uriLen && *pURILoc )
+        {
+            if( isalnum( *pURILoc ) || ( *pURILoc == '-' ) || ( *pURILoc == '_' ) || ( *pURILoc == '.' ) || ( *pURILoc == '~' ) )
+            {
+                *pBufLoc++ = *pURILoc;
+                index++;
+            }
+            else if( ( *pURILoc == '/' ) && !encodeSlash )
+            {
+                *pBufLoc++ = *pURILoc;
+                index++;
+            }
+            else
+            {
+                *pBufLoc++ = '%';
+                *pBufLoc++ = intToHex( *pURILoc >> 4 );
+                *pBufLoc++ = intToHex( *pURILoc & 15 );
+
+                index += 3;
+            }
+
+            pURILoc++;
+        }
+
+        if( nullTerminate )
+        {
+            *pBufLoc++ = '\0';
+            index++;
+        }
+
+        *canonicalURILen = index;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static void generateCanonicalURI( const char * pURI,
+                                      size_t uriLen,
+                                      bool encodeOnce,
+                                      canonicalContext_t * canonicalRequest )
+    {
+        size_t encodedLen, remainingLen = canonicalRequest->bufRemaining;
+        char * pBufLoc = canonicalRequest->pBufCur;
+
+        assert( pURI != NULL );
+        assert( canonicalRequest != NULL );
+
+        encodeURI( pURI, uriLen, pBufLoc, &encodedLen, false, true );
+
+        remainingLen -= encodedLen;
+
+        if( !encodeOnce )
+        {
+            encodeURI( pBufLoc, encodedLen, pBufLoc + encodedLen, &remainingLen, false, true );
+            memmove( canonicalRequest->pBufCur + encodedLen, canonicalRequest->pBufCur, remainingLen );
+        }
+
+        canonicalRequest->pBufCur += remainingLen;
+        *( canonicalRequest->pBufCur++ ) = '\n';
+
+        canonicalRequest->bufRemaining -= remainingLen + 1;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static void generateCanonicalQuery( const char * pQuery,
```

Comment:
It will be better if more comments can be added to the function to understand it easily.